text/0000-native-class-constructor-update.md

```diff
@@ -0,0 +1,228 @@
+behavior.
+
+## Detailed design
+
+One very important design constraint to making this change is that we _cannot_
+break the behavior of EmberObject when used _without_ native classes. To do
+this, we will leverage the fact that the static `create` method is the only
+public way to create an instance of EmberObject.
+
+Currently, the behavior of EmberObject is the following (simplified):
+
+```js
+class EmberObject {
+  constructor(props) {
+    // ..class setup things
+
+    Object.assign(this, props);
+    this.init();
+  }
+
+  static create(props) {
+    let instance = new this(props);
+
+    return instance;
+  }
+}
+```
+
+We can change it to the following (simplified):
+
+```js
+class EmberObject {
+  constructor(props) {
+    // ..class setup things
+  }
+
+  static create(props) {
+    let instance = new this(props);
+
+    Object.assign(this, props);
+    this.init();
+
+    return instance;
+  }
+}
+```
+
+This would assign the properties _after_ all of the class fields for any
+subclasses have been assigned. Revisiting our previous example, the following
+two class declarations would effectively be equivalent:
+
+```js
+const Foo = EmberObject.extend({
+  bar: 'baz'
+});
+
+class Foo extends EmberObject {
+  bar = 'baz';
+}
+```
+
+Much easier to codemod! There are other subtle differences between native class
+fields and EmberObject properties, such as the fact that class fields are
+assigned each time a class is initialized, but these are easier to work around.
+
+### Injections and the `init` hook
+
+One side effect of this change is that injections will not be available on the
+class instance during the `constructor` phase. This behavior is not very
+commonly used - based on an informal community survey we found only a few usages
+\- but it _does_ exist and have its use cases.
+
+Figuring out the ideal behavior of injections during the constructor phase is
+outside of the scope of this RFC, and is something that should be discussed in
+future RFCs. For the time being, users can still rely on the `init` hook, which
+will continue to be called after all injections and properties have been
+assigned to the instance.
+
+### `new EmberObject()`
+
+It was previously possible to use `new` syntax with EmberObject. While this
+was not considered public API, it has technically worked and been under test
+since the early days of Ember, and may fall under the category of intimate API.
+
+We can continue to support this behavior in a backwards compatible way while
+deprecating it with one final tweak to the change above:
+
+```js
+const DEFER_INIT = new Symbol();
+
+function initialize(instance, props) {
+  Object.assign(instance, props);
+  instance.init();
+}
+
+class EmberObject {
+  constructor(props, maybeDefer) {
+    // ..class setup things
+
+    if (maybeDefer === DEFER_INIT) {
+       return this;
+    }
+
+    deprecate('using new with EmberObject has been deprecated', true);
```

Comment:
I'd like to see this fleshed out to include the deprecation timeline, what would the `until` be here? I'd _personally_ like to see this be considered a private API deprecation, where `until` would be the next LTS+1...

Reply:
I'd also support that timeline.

Reply:
I think the sooner we can remove it, the better. Will update the RFC to state that it will be `until: '3.5.0'`